### ISSUE TITLE
window position is always saved and restored, even when `let g:neovide_remember_window_position=v:false`

### ISSUE THREAD

**Issue by correabuscar** [neovide/neovide#1580]
Neovide always saves window position and thus (since it's there, saved) it always restores it!
Saves it as `0,0` when `let g:neovide_remember_window_position = v:false`.

ie. 
`DEBUG [neovide::settings::window_geometry] Saved Window Settings: {"window":{"Windowed":{"position":{"x":0,"y":0},"size":{"width":100,"height":50}}}}`
`DEBUG [neovide::settings::window_geometry] Loaded window settings: Windowed { position: PhysicalPosition { x: 0, y: 0 }, size: Dimensions { width: 100, height: 50 } }`

Workaround patch (ie. don't always restore position):
```patch
diff --git a/src/window/mod.rs b/src/window/mod.rs
index 0dccb94..8f74de3 100644
--- a/src/window/mod.rs
+++ b/src/window/mod.rs
@@ -295,7 +295,15 @@ pub fn create_window() {
                 maximized = true;
             }
             PersistentWindowSettings::Windowed { position, .. } => {
-                previous_position = Some(position);
+                // don't position window to 0,0 when: let g:neovide_remember_window_position = v:false
+                // but rather, let user's window manager position it
+                let remember_window_position = { SETTINGS.get::<WindowSettings>().remember_window_position };
+                if remember_window_position {
+                    // ugly hack because position is always saved: as 0,0 when this is false
+                    // FIXME: don't save window position when remember_window_position is false!
+                    // seen in settings/window_geometry.rs
+                    previous_position = Some(position);
+                }
             }
         }
     }
```

Fix: TODO: don't [always save position](https://github.com/neovide/neovide/blob/0b6e20c1994c761814eb6a8ad33dea90ade97573/src/settings/window_geometry.rs#L93-L98) ? but rather only when `remember_window_position` is true.

Side fix (unrelated): `repositioned window: true` was always printed even if it were `false`
```patch
diff --git a/src/window/mod.rs b/src/window/mod.rs
index 0dccb94..8f74de3 100644
--- a/src/window/mod.rs
+++ b/src/window/mod.rs
@@ -376,9 +384,10 @@ pub fn create_window() {
                 || window_position.y > monitor_position.y + monitor_height
             {
                 window.set_outer_position(monitor_position);
+                Some(())
+            } else {
+                None
             }
-
-            Some(())
         })
         .is_some();
 
```
now: `TRACE [neovide::window] repositioned window: false`

**Comment by correabuscar** [comment]
The above patch doesn't work when `neovide` was compiled with `cargo install  --path .` ie. without `--debug`, because for some reason (not known yet) the [default](https://github.com/neovide/neovide/blob/0b6e20c1994c761814eb6a8ad33dea90ade97573/src/window/settings.rs#L30) (true) value of `remember_window_position` is used when `neovide` is compiled as release, instead of the value false ie. `let g:neovide_remember_window_position = v:false
` from `~/.config/nvim/init.vim`. I'm trying to figure out why...

**Comment by 9mm** [comment]
where is "neovide_remember_window_position" coming from? all i see is remembering window_size

Im trying to get neovide to remember both size AND position when I re-open, it seems everything I try neovide 100% ignores it and always shows the identical way.

ive tried this neovide_remember_window_size as well as passing `--geometry 500x500` to the bin and it still is ignored

**Comment by fredizzimo** [comment]
The best fix for this is probably to change the following function to load the existing config first, and only change the fields that the user has requested to update.

https://github.com/neovide/neovide/blob/39af18d647b4aeff43d6829f43fe4e57628691c0/src/settings/window_size.rs#L64-L93

**Comment by 9mm** [comment]
@fredizzimo can you summarize how I might add PR's to neovide? I've never used rust, i dont need a lengthy explanation maybe just like 3-4 cliffnotes. Does this sound right?

1. git clone repo
2. make changes
3. `cargo build`
4. see if problem goes away
5. create PR

my questions:

1. do I have to run tests or do anything else?
2. can i incrementally compile somehow?

Thanks

**Comment by fredizzimo** [comment]
@9mm,

The request workflow is the same as basicalky anything else on github, so create a branch on your fork and create a pr from there.

The CI here is configured a bit differently than for most other projects, the full test suite will run on your fork so make sure it passes there first.

Locally you can run:
`cargo test` for unit tests
`cargo clippy` for static analyzing
`cargo fmt` for code formatting
`cargo run` to build and run, add `—release` for a release build instead of debug

Make sure that all those pass locally first

The compilation is incremental by default.

**Comment by 9mm** [comment]
awesome, thanks!

**Comment by 9mm** [comment]
@fredizzimo maybe I'll give this a try again.

I started at your description for some time:

```
The best fix for this is probably to change the following function to load the existing config first, and only change the fields that the user has requested to update.
```

Could extrapolate just a little more of what the race condition in this bit of code is stemming from?

From what I can tell it...

1. takes maximized, window size, and position as inputs
2. sets a windows setting object which includes the vim configuration settings
3. builds a PersistentWindowSettings object based on current window settings + vim settings
4. serializes the PersistentWindowSettings object

**Comment by fredizzimo** [comment]
@9mm, at least the original request was to not update the windows position when `let g:neovide_remember_window_position = v:false`

That means that we have to ensure that the old value remains in the settings when saving. Currently it always creates an entirely new configuration, which means that the windows position and everything else that is not saved ends up being reset to the default values, rather than using the value that was saved.

So that's why I'm suggesting to first read the old configuration, and then update the values that are set to save based on the current runtime values, and then write out everything to the file.

NOTE1: that I don't think there is any race condition in the current version of the code, but there might have been in the past. 

NOTE2: I can imagine that some users expect the value to be reset to the default when not saved. But I think for that use case, we should add command-line arguments to reset the settings. A workaround for resetting them currently is to first have remember settings on, then pass the desired size and position on the command line and exit. Which should save that position and size.

**Comment by 9mm** [comment]
@fredizzimo Ah ok, so I think there might be a misunderstanding. My issue is actually the opposite of his :)

In this issue above, it seems neovim ignores his `remember_window_size ` setting and just *always* restores it.

In my case, Neovim will remember the window position about half the time (the desired outcome) but the other half it will default to 1/4 of my screen in the top right.

I think the original thing that made me reference this issue from my [other issue](https://github.com/neovide/neovide/issues/1951) was the fact he mentioned:

```
thanks to cehteh on libera chat #rust channel for suggesting a delay, I can now confirm that the nvim settings are loaded in parallel with the reading of the settings, thus anything like let g:neovide_transparency=0.8 being applied in time is just a coincidence
```

Which seems to indicate some kind of race even if his problem is the opposite.

I would imagine it could be related to this other one:

https://github.com/neovide/neovide/issues/1590

**Comment by fredizzimo** [comment]
@9mm, what you are seeing is most likely has the same underlying cause as this https://github.com/neovide/neovide/issues/1972. 

I think the settings are both saved and loaded correctly, but if we send the resize command at the wrong time, it's ignored on macOS. 

I have been looking at creating the window with the right size from the start, so we don't need to resize it, and I think that will work, except that this becomes hard to solve https://github.com/neovide/neovide/issues/620.

**Comment by 9mm** [comment]
Very interesting... I also have paddings set for sure:

```lua
--                          __
-- .-----.-----.-----.--.--|__.--------.
-- |     |  -__|  _  |  |  |  |        |
-- |__|__|_____|_____|\___/|__|__|__|__|

-- essentials
vim.g.mapleader = ','
vim.g.maplocalleader = ','
vim.opt.termguicolors = true

-- lazy
local lazypath = vim.fn.stdpath('data') .. '/lazy/lazy.nvim'
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({
    'git',
    'clone',
    '--filter=blob:none',
    'https://github.com/folke/lazy.nvim.git',
    '--branch=stable',
    lazypath,
  })
end
vim.opt.rtp:prepend(lazypath)
require('lazy').setup('zesty.plugins', {
  change_detection = {
    enabled = true,
    notify = false,
  },
})

-- gui
require('zesty.gui')

-- keymaps
require('zesty.keymaps')

-- autocommands
require('zesty.autocommands')

-- aesthetic
vim.o.guifont = 'AnonymicePro Nerd Font:h15'

-- indentation
vim.opt.autoindent  = true -- continue indentation to new line
vim.opt.smartindent = true -- add extra indent when it makes sense
vim.opt.smarttab    = true -- <tab> at the start of a line behaves as expected
vim.opt.expandtab   = true -- <tab> inserts spaces
vim.opt.shiftwidth  = 2    -- >>, << shift line by 4 spaces
vim.opt.tabstop     = 2    -- <tab> appears as 4 spaces
vim.opt.softtabstop = 2    -- <tab> behaves as 4 spaces when editing

-- cursor
--vim.opt.guicursor = ''
vim.opt.virtualedit = 'all'

-- line numbers
vim.opt.number = true
vim.opt.signcolumn = 'yes:1' -- auto:1'

-- wrap
vim.opt.wrap = false
vim.opt.linebreak = true
vim.opt.breakindent = true
--vim.o.completeopt = 'menuone,noselect'

-- scroll
vim.opt.scrolloff = 4
vim.opt.mouse = 'nv'
vim.opt.mousescroll = 'ver:2,hor:0'

-- splits
vim.opt.splitright = true
vim.opt.splitbelow = true

-- search
vim.opt.hlsearch = true
vim.opt.incsearch = true
vim.opt.ignorecase = true
vim.opt.smartcase = true

-- recovery
vim.opt.backup = false
vim.opt.undolevels = 1000
vim.opt.undofile = false
vim.opt.undodir = os.getenv('HOME') .. '/.local/state/nvim/undo//'
vim.opt.swapfile = true
vim.opt.directory = os.getenv('HOME') .. '/.local/state/nvim/swap//'

-- misc
vim.opt.updatetime = 50
vim.o.timeout = true
vim.o.timeoutlen = 500

if vim.g.neovide then

  local function close_tab()
    local bufnr = vim.api.nvim_get_current_buf()
    local modified = vim.api.nvim_get_option_value('modified', { buf = bufnr })
    if modified then
      vim.ui.input(
        { prompt = '... unwritten buffer, are you sure? (y/n) ' },
        function(input)
          if input == 'y' then
            vim.cmd('bd!')
          end
        end
      )
    else
      vim.cmd('bd')
    end
    vim.cmd("echo ''")
  end

  local alpha = function()
    return string.format('%x', math.floor(255 * (vim.g.transparency or 0.8)))
  end

  -- padding
  vim.g.neovide_padding_top    = 20
  vim.g.neovide_padding_bottom = 20
  vim.g.neovide_padding_right  = 20
  vim.g.neovide_padding_left   = 20

  -- display
  vim.g.neovide_transparency = 0.0 -- must be 0.0 for menu bar to match
  vim.g.transparency = 0.98
  vim.g.neovide_background_color = '#0f0f12' .. alpha() -- match theme.lua
  vim.g.neovide_refresh_rate = 120
  vim.g.neovide_refresh_rate_idle = 5
  vim.g.neovide_hide_mouse_when_typing = true
  vim.g.neovide_underline_automatic_scaling = true

  -- window
  vim.g.neovide_remember_window_size = true
  vim.g.neovide_confirm_quit = true

  -- animation
  vim.g.neovide_cursor_animation_length = 0 -- 0.06
  vim.g.neovide_cursor_trail_size = 0 -- 0.4
  vim.g.neovide_cursor_animate_command_line = false
  vim.g.neovide_scroll_animation_length = 0.3

  -- keys
  vim.g.neovide_input_macos_alt_is_meta = true

  -- command mapping
  vim.keymap.set({ 'i', 'n' }, '<D-a>', '<ESC>ggVG')                              -- select all
  vim.keymap.set({ 'i', 'n' }, '<D-w>', function() close_tab() end)               -- close tab
  vim.keymap.set({ 'i', 'n' }, '<D-[>', function() vim.cmd('BufferPrevious') end) -- previous tab
  vim.keymap.set({ 'i', 'n' }, '<D-]>', function() vim.cmd('BufferNext') end)     -- next tab
  vim.keymap.set('i', '<D-t>', '<C-o>:tabnew<CR><ESC>')                           -- new tab (insert)
  vim.keymap.set('n', '<D-t>', ':tabnew<CR>')                                     -- new tab (insert)
  vim.keymap.set('i', '<D-s>', '<C-o>:w<CR>')                                     -- save (insert)
  vim.keymap.set('n', '<D-s>', ':w<CR>')                                          -- save (normal)
  vim.keymap.set('x', '<D-x>', '"+d')                                             -- cut
  vim.keymap.set('x', '<D-c>', '"+y')                                             -- copy
  vim.keymap.set('i', '<D-v>', '<C-r><C-o>+')                                     -- paste (insert)
  vim.keymap.set('n', '<D-v>', 'i<C-r><C-o>+<ESC>l')                              -- paste (normal)
  vim.keymap.set('x', '<D-v>', '"+P')                                             -- paste (visual)
  vim.keymap.set('c', '<D-v>', '<C-r>+')                                          -- paste (command)
  -- vim.keymap.set('n', '<D-q>', ':q<CR>')                                          -- quit

end

```
